user/tinytcl/extensions/syslog/syslog.man
syslog(n)		     Tcl Syslog extension		     syslog(n)



______________________________________________________________________________

NAME
       syslog - send message to syslog from Tcl script

SYNOPSIS
       load syslog.so syslog

       syslog ?options? ?priority? message
_________________________________________________________________


DESCRIPTION
       This  command sends message to system syslog facility with given prior-
       ity. Valid priorities are:

       emerg, alert, crit, err, error, warning, notice, info, debug.

       If a message is specified, but no priority is specified, then a	prior-
       ity of info is used.

       By  default, facility user is used and the value of global tcl variable
       argv0 is used as ident string.  However, any of the  following  options
       may be specified before priority to control these parameters:

       -facility value	   Use	specified  facility instead of user. Following
			   facility are recognized:

			   authpriv, cron, daemon, kernel, lpr,
			   mail, news, syslog, user, uucp, local0-local7.


       -ident string	   Use given string  instead  of  argv0	 variable  for
			   ident string.

       -options integer	   Set syslog options such as LOG_CONS, LOG_NDELAY You
			   should use numeric values of those from your system
			   syslog.h file, becouse I haven't got time to imple-
			   ment yet another hash table.



KEYWORDS
       logging, syslog



Tcl				      2.0			     syslog(n)


freeswan/CHANGES
New and notable in 1.97:

Sanity check added for at least one of AH or ESP more for checking if
"make *config" has even been run.

Fixed 2.2 local IKE fragmentation blackhole.  Still won't work if
iptraf or another pcap app is running.

MCR added ikeping(8).

Applied DHR's tunnel patch to streamline IKE/specialSA processing.

Added local and remote IKE bypass tests to the UML test suite.

UML KLIPS %trap* tests now check for upbound pf_key messages.

An ICMP ICMP_DEST_UNREACH, ICMP_PKT_FILTERED is now sent upon %reject.

Added a UML input file with first and last packets being outside the 
covered range to "sandwich" the packets under test.

The default _updown script has been cleaned up slightly, reducing
internal duplication that arose accidentally.

When doing --add on an RSA-signature conn, ipsec_auto now tells Pluto
about the keys before, not after, telling it about the conn.  The main
consequence of this is that if something is wrong with one of the keys,
the conn itself does *not* get loaded.  This should reduce confusion.

The rp_filter checks during KLIPS startup have been fixed to be a bit less
paranoid; rp_filter is thought to cause problems only on the physical
interfaces, not on the ipsecN interfaces. 

The configuration process has been fixed to be a bit more paranoid.  Some
users have had mysterious compile failures with complaints involving names
like MD5_CTX; they seem to have resulted from kernel configuration changes
somehow getting into .config but not into include/linux/autoconf.h (which
is what actually controls most of the kernel compile).  The post-config
checking now verifies not only that .config has the right stuff in it, but
that autoconf.h does too.  (We still don't see *how* this discrepancy can
arise, but this will at least catch it when it does.)

The details of linking the KLIPS source into the kernel are now under
control of a configuration variable in Makefile.inc, to make it easier
for distribution builders to change.

The top-level Makefile now includes cleverness which assigns a date-based
version code if a build is done from a raw CVS checkout.

The library now includes a new function, ttodatav(), like ttodata() but
in some cases it gives more verbose error reports.  Also, there are now
a couple of functions which distill an RSA public key to a "key ID" for
use in messages, a crypto-quality pseudo-random-number generator, and a
function that converts a single address into a (singleton) subnet.

Barf's secret/key-censoring utilities have been tidied up so that what
they emit matches the key-ID library functions. 

The ipsec command now supplies a couple more environment variables to
things it calls, and this has reduced the number of things which need to
be edited at build time to know about configuration variables. 

The configuration reader now allows parameter names to begin with "_";
these are strictly reserved for internal development purposes and will
not in general be documented.


As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.96:

Compatibility with the 2.0.xx kernels (under Red Hat 5.2) is broken.
This was accidental, and we have some hopes of fixing it, although it
won't last forever even if we do.  See README.

Our "IKE Implementation Issues" IETF draft has been added to our docs.

Newhostkey now has a mandatory --output "option" specifying a filename
(although the magic filename "-" means "standard output"), and sets
restrictive permissions on newly-created files to try to keep private keys
truly private. 

Pluto now gracefully handles a couple of odd error cases (missing private
keys, network errors at just the wrong time in negotiations) which used to
cause it to die, provoking an automatic restart of the IPsec subsystem. 

Makefile.inc now has more detailed control over where manpages go, to ease
integration into distributions.  The name of the destination directory for
RPMs is also now centralized there.

Rsasigkey now checks for certain obscure internal failures, which can
happen if --random is used to supply a source of random bits that aren't
really very random, and complains accordingly.

Barf's eroute table, formerly sorted by source, is now sorted by destination.
This is probably a bit clearer for hosts with lots of eroutes.

Use of the %reject target in a shunt eroute now, as long planned, actually
causes an ICMP rejection message (Destination Unreachable:  Communication
Administratively Prohibited). 

KLIPS now declares its module license to be "Dual BSD/GPL".  (It is 
dual because the libdes and zlib code is BSD licensed, while KLIPS itself 
is GPL).  Whether this is exactly right is not yet clear, but at least it
stops the latest module loaders from whining.

The UML test stuff continues to grow and add new test cases.

As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.95:

The last remnants of the "%hold" bug, which broke 1.93 and 1.94, have
(we think) been dealt with.

Minor bugs in the RPM-building stuff have been fixed, and it has been
updated to deal with Red Hat's abrupt removal of the --buildroot option
from rpm.

Pluto now supports the ability to delete an instance of a conn, rather
than just the whole conn (for cases, notably OE, where a single connection
description can be involved in multiple connections).  There is no support
for this in ipsec_auto yet.

Pluto's routine (non-debug) logging is now somewhat terser, and some of
the more boring messages are produced only if --debug-lifecycle is turned
on (this is a temporary hack, it's not really debugging output).

Pluto's innards have been reorganized somewhat, in preparation for making
DNS lookups asynchronous.  There should be no user-visible effects.

Pluto now ignores IPv6 link-local interface addresses; we think that they
are never relevant, and trying to bind to them causes difficulties.  This

user/tinytcl/extensions/syslog/syslog.c
/* Syslog interface for tcl
 * Copyright Victor Wagner <vitus@ice.ru> at
 * http://www.ice.ru/~vitus/works/tcl.html#syslog
 *
 * Slightly modified by Steve Bennett <steveb@snapgear.com>
 */
#include <tcl.h>
#include <tclInt.h>
#include <syslog.h>
#include <string.h>
typedef struct {
                int logOpened;
                long facility;
				int options;
                char ident[32];
                Tcl_HashTable *priorities;
                Tcl_HashTable *facilities;
               } SyslogInfo;

#define Tcl_Alloc ckalloc
#define Tcl_Free ckfree

static void Syslog_ListHash(Tcl_Interp *interp,Tcl_HashTable *table);         
/* SyslogHelp - puts usage message into interp->result
 * 
 *
 */

static void SyslogHelp(Tcl_Interp *interp,char *cmdname)
{  Tcl_AppendResult(interp,"Wrong # of args. should be ",cmdname, 
           " ?option value? ?priority? message",NULL);
}

/* Syslog_Log -
 * implements syslog tcl command. General format: syslog ?options? level text
 * options -facility -ident -options
 * 
 */
static int Syslog_Log(ClientData data, Tcl_Interp *interp, int argc, char **argv)
{    SyslogInfo *info=(SyslogInfo *)data;
    char *message = NULL;
    long priority = LOG_INFO;
    int i=1;
    if (argc<=1) {
        SyslogHelp(interp,argv[0]);
        return TCL_ERROR;
    }
  while (i<argc-1) {
    if (!strcmp(argv[i],"-facility")) {
        Tcl_HashEntry * entry=Tcl_FindHashEntry(info->facilities,argv[i+1]);
        if (!entry) {
           Tcl_AppendResult(interp,"Invalid facility name: \"",argv[i+1],
                   "\" available facilities:",
               NULL);
           Syslog_ListHash(interp,info->facilities);
           return TCL_ERROR;
        }
        info->facility=(long)Tcl_GetHashValue(entry);
        if (info-> logOpened) {
            closelog();
            info-> logOpened=0;
        }
     } else if (!strcmp(argv[i],"-options")) {
         int tmp;
        if (Tcl_GetInt(interp,argv[i+1],&tmp)==TCL_ERROR)
             return TCL_ERROR;
        info->options=tmp;
        if (info->logOpened) {
            closelog();
            info->logOpened=0;
        }
     } else if (!strcmp(argv[i],"-ident")) {
        strncpy(info->ident, argv[i+1],32);
        info->ident[31]=0;
        if (info->logOpened) {
            closelog();
            info->logOpened=0;
        }
     } else {
       break;
     }
     i+=2;
  }

  /* There should be either 0, 1 or 2 args left */
  if (i == argc) {
    /* No args, but they have set some options, so OK */
    return TCL_OK;
  }

  if (i<argc-1) {
       Tcl_HashEntry *entry=Tcl_FindHashEntry(info->priorities,argv[i]);
       if (!entry) {
          Tcl_AppendResult(interp,"Invalid syslog level \"",argv[i],
                  "\" available levels:",
               NULL);
          Syslog_ListHash(interp,info->priorities); 
          return TCL_ERROR;
       }
       priority=(long)Tcl_GetHashValue(entry);
       i++;
  }
  message=argv[i];
  i++;

  if (i != argc) {
      SyslogHelp(interp,argv[0]);
      return TCL_ERROR;
  }
  if (!info->logOpened) {
      openlog(info->ident,info->options,info->facility);
      info->logOpened=1;
  }
  syslog(priority,"%s",message);
  return TCL_OK;
}

/*
 * Syslog_ListHash - appends to interp result all the values of given
 * hash table
 */
static void Syslog_ListHash(Tcl_Interp *interp,Tcl_HashTable *table) 
{
    Tcl_HashSearch *searchPtr=(Tcl_HashSearch *)
          Tcl_Alloc(sizeof(Tcl_HashSearch));
    Tcl_HashEntry *entry;
    char separator[3]={' ',' ',0};   
    entry=Tcl_FirstHashEntry(table,searchPtr);
    while (entry) {
        Tcl_AppendResult(interp,separator,Tcl_GetHashKey(table,entry),NULL);
        separator[0]=',';
        entry=Tcl_NextHashEntry(searchPtr);
    }   
    Tcl_Free((char *)searchPtr);
} 

/* 
 *  Syslog_Delete - Tcl_CmdDeleteProc for syslog command.
 *  Frees all hash tables and closes log if it was opened.
 */
static void Syslog_Delete(ClientData data)
{ SyslogInfo *info=(SyslogInfo *)data;
  Tcl_DeleteHashTable(info->facilities);
  Tcl_Free((char *)info->facilities);
  Tcl_DeleteHashTable(info->priorities);
  Tcl_Free((char *)info->priorities);
  if (info->logOpened) {
     closelog();
  }
  Tcl_Free((char *)info);
}
/*
 * My simplified wrapper for add values into hash
 *
 */
static void AddEntry(Tcl_HashTable *table,char *key,int value)
{ int new;
  Tcl_HashEntry *entry=Tcl_CreateHashEntry(table,key,&new);
  Tcl_SetHashValue(entry,(ClientData)((long)value));
}

/*
 * Syslog_Init 
 * Package initialization procedure for Syslog package. 
 * Creates command 'syslog', fills hash tables to map symbolic prioriry 
 * and facility names to system constants.
 */
int syslog_Init(Tcl_Interp *interp)
{  char *argv0;
   SyslogInfo *info=(SyslogInfo *)Tcl_Alloc(sizeof(SyslogInfo));
   info->logOpened=0;
   info->options=0;
   info->facility=LOG_USER;
   argv0=Tcl_GetVar(interp,"argv0",TCL_GLOBAL_ONLY);
   if (argv0) {
       strncpy(info->ident,argv0,32);
   } else {
       strcpy(info->ident,"Tcl script");
   }
   info->ident[31]=0;
   info->facilities =(Tcl_HashTable *) Tcl_Alloc(sizeof(Tcl_HashTable));
   Tcl_InitHashTable(info->facilities,TCL_STRING_KEYS);
   AddEntry(info->facilities,"auth",LOG_AUTH);  
#ifndef LOG_AUTHPRIV
# define LOG_AUTHPRIV LOG_AUTH
#endif
   AddEntry(info->facilities,"authpriv",LOG_AUTHPRIV);  
   AddEntry(info->facilities,"cron",LOG_CRON);  
   AddEntry(info->facilities,"daemon",LOG_DAEMON);  
   AddEntry(info->facilities,"kernel",LOG_KERN);
   AddEntry(info->facilities,"lpr",LOG_LPR);
   AddEntry(info->facilities,"mail",LOG_MAIL);
   AddEntry(info->facilities,"news",LOG_NEWS);
   AddEntry(info->facilities,"syslog",LOG_SYSLOG);
   AddEntry(info->facilities,"user",LOG_USER);
   AddEntry(info->facilities,"uucp",LOG_UUCP);
   AddEntry(info->facilities,"local0",LOG_LOCAL0);
   AddEntry(info->facilities,"local1",LOG_LOCAL1);
   AddEntry(info->facilities,"local2",LOG_LOCAL2);
   AddEntry(info->facilities,"local3",LOG_LOCAL3);
   AddEntry(info->facilities,"local4",LOG_LOCAL4);
   AddEntry(info->facilities,"local5",LOG_LOCAL5);
   AddEntry(info->facilities,"local6",LOG_LOCAL6);
   AddEntry(info->facilities,"local7",LOG_LOCAL7);
   info->priorities = (Tcl_HashTable *) Tcl_Alloc(sizeof(Tcl_HashTable));
   Tcl_InitHashTable(info->priorities,TCL_STRING_KEYS);
   AddEntry(info->priorities,"emerg",LOG_EMERG);
   AddEntry(info->priorities,"alert",LOG_ALERT);
   AddEntry(info->priorities,"crit",LOG_CRIT);
   AddEntry(info->priorities,"err",LOG_ERR);
   AddEntry(info->priorities,"error",LOG_ERR);
   AddEntry(info->priorities,"warning",LOG_WARNING);
   AddEntry(info->priorities,"notice",LOG_NOTICE);
   AddEntry(info->priorities,"info",LOG_INFO);
   AddEntry(info->priorities,"debug",LOG_DEBUG);
   Tcl_CreateCommand(interp,"syslog",Syslog_Log,(ClientData) info,
            Syslog_Delete); 
   return TCL_OK;
}
